### PR TITLE
fix #3960: adding a KMS helper method and additional mock crd support

### LINKED ISSUE #3960: Add mock handling for implicit generic support
### Is your task related to a problem? Please describe

When using the built-in mock server the metadata will be missing for a generic resource.  

### Describe the solution you'd like

There should be a way either against the client or the mock server to register a ResourceDefinitionContext for testing purposes.

Currently when you use the entry point client.genericKubernetesResources(context) - that context is not cached / saved, it is only used locally for that operation.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shawkins**
## Description
This furthers crd support for the KubernetesMockServer.  If you are in crud mode and manipulate CRDs, that will create the appropriate apis/group/version metadata.

There is also a KubernetesMockServer.expectCustomResource method that does the same thing, but starting from a CustomResouceDefinitionContext - however the behavior of crud and non-crud is slightly different.

From https://github.com/fabric8io/kubernetes-client/issues/3960#issuecomment-1192485822 this addresses 2 - installing crds.  It does not address 3 in a higher-level way - you would need to use a setup method to configure which crds are expected. 

4 - not supporting custom resources that have not been installed, is only how we behave when using the genericKubernetesResources(apiVersion, kind) entry point.  Any other usage of a custom resource is allowed in crud mode.  For it to be disallowed we'd have to delineate between custom and non-custom resources, which needs a bit more discussion.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/CustomResourceDefinitionProcessor.java [hunk 1]
```diff
@@ -49,19 +58,80 @@ public void process(String path, String crdString, boolean delete) {
...
+    return Optional.ofNullable(this.crdContexts.computeIfPresent(key(context), (k, v) -> {
+      v.remove(context.getPlural());
+      return v;
+    })).map(Map::values);
+  }
+
+  /**
+   * Remove the given context and return a collection of the remaining contexts for the api/version
+   */
+  public Optional<Collection<CustomResourceDefinitionContext>> addCrdContext(CustomResourceDefinitionContext context) {
```
Comment (rohanKanojia) [inline junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/CustomResourceDefinitionProcessor.java:103]: nit, maybe javadoc doesn't match with what method is doing?

File: kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/KubernetesMockServerTest.java [hunk 2]
```diff
@@ -17,18 +17,28 @@
...
 import io.fabric8.openshift.client.OpenShiftClient;
 import org.junit.jupiter.api.Test;
 
 import static org.junit.Assert.assertNull;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @EnableKubernetesMockClient(crud = true)
```
Comment (rohanKanojia) [inline kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/KubernetesMockServerTest.java:44]: Not directly related to this PR but this test class name looks a bit misleading. All tests in this module are KubernetsMockServer tests.

**Review by rohanKanojia — APPROVED** [review]

File: junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/CustomResourceDefinitionProcessor.java (hunk 1 above)
Comment (shawkins) [reply junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/CustomResourceDefinitionProcessor.java]: Good catch.  It's now been simplified in general.

File: junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesMockServer.java [hunk 3]
```diff
@@ -188,4 +195,39 @@ public void reset() {
...
+   * <p>
+   * If this is a crud server, the custom resource will be added to the set of previously added resources
+   * and the resources inferred from custom resource definitions that have been added.
+   * <p>
+   * If this server is not crud, this call will add a single expectation for the given resource. Direct handling of
+   * multiple resources for a given api group/version has not yet been added.
+   *
+   * @param rdc the resource definition context
+   */
+  public void expectCustomResource(CustomResourceDefinitionContext rdc) {
```
Comment (manusa) [inline junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesMockServer.java:214]: Quick question, why doesn't this class implement the new `CustomResourceAware` interface?

**Review by manusa — APPROVED** [review]
LGTM

File: junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesMockServer.java (hunk 3 above)
Comment (shawkins) [reply junit/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesMockServer.java:215]: No reason, I've added that now.

**Review by manusa — APPROVED** [review]

**Pull request merged**